File: apps/server/dekobloko_server/huffman.py
"""Chat text Huffman codec.

Chat bodies are compressed with a table the client loads from archive 3, file
"huffman" (client.java:4669: `new jk(cl.field_y.a(0, "huffman", ""))`). The raw
table is 256 bytes -- one code length per character -- but the code VALUES are
produced by jk's constructor, which assigns them in character order with
per-length counters and carry propagation between lengths.

That construction was reimplemented by hand twice and was wrong both times
(plain canonical ordering yields 'sfee'/'yb' for known-good input; a literal
translation of the constructor yields nothing at all). Rather than keep
guessing, the code table was dumped from the client's OWN jk class -- loaded in
a JVM with the real table bytes and read back by reflection -- into
huffman-codes.csv as `char,bitlength,code` rows.

Verified against captured traffic: `e7 bc` -> "test", `8d 09 80` -> "lmao".

If the cache is ever replaced, regenerate the CSV rather than editing it; the
values are derived, not authored.
"""
from __future__ import annotations

from pathlib import Path

_CSV = Path(__file__).resolve().parent / "huffman-codes.csv"
_table: dict[tuple[int, int], int] | None = None


def _load() -> dict[tuple[int, int], int]:
    global _table
    if _table is None:
        rows = (
            line.split(",")
            for line in _CSV.read_text().splitlines()
            if line.strip()
        )
        _table = {(int(ln), int(code)): int(ch) for ch, ln, code in rows}
    return _table


def decode(data: bytes, count: int) -> str:
    """Decode `count` characters from Huffman-compressed `data`.

    Bits are consumed MSB-first. `count` comes from the packet (li.a caps it at
    80); decoding stops there even if bits remain, which is expected since the
    final byte is usually padded.
    """
    table = _load()
    out: list[str] = []
    cur = 0
    length = 0
    for byte in data:
        for bit in range(7, -1, -1):
            cur = (cur << 1) | ((byte >> bit) & 1)
            length += 1
            if (length, cur) in table:
                out.append(chr(table[(length, cur)]))
                cur = 0
                length = 0
                if len(out) == count:
                    return "".join(out)
    return "".join(out)

File: apps/server-js/src/huffman.js
'use strict';

// Port of dekobloko_server/huffman.py: chat text Huffman codec.
//
// Chat bodies are compressed with the table the client loads from archive 3,
// file "huffman". Code VALUES were dumped from the client's own jk class into
// huffman-codes.csv (char,bitlength,code rows) rather than reimplemented.
// Verified against captured traffic: e7 bc -> "test", 8d 09 80 -> "lmao".

const fs = require('fs');
const path = require('path');

const CSV_PATH = path.join(__dirname, 'huffman-codes.csv');
let _table = null;

function load() {
  if (_table === null) {
    _table = {};
    const rows = fs
      .readFileSync(CSV_PATH, 'utf8')
      .split(/\r?\n/)
      .filter((line) => line.trim() !== '');
    for (const row of rows) {
      const [ch, ln, code] = row.split(',');
      const bitLength = parseInt(ln, 10);
      const codeValue = parseInt(code, 10);
      if (!_table[bitLength]) _table[bitLength] = {};
      _table[bitLength][codeValue] = parseInt(ch, 10);
    }
  }
  return _table;
}

// Decode `count` characters from Huffman-compressed `data`.
// Bits are consumed MSB-first; decoding stops at count even if bits remain,
// which is expected since the final byte is usually padded.
function decode(data, count) {
  const table = load();
  const out = [];
  let cur = 0;
  let length = 0;
  for (const byte of data) {
    for (let bit = 7; bit >= 0; bit -= 1) {
      cur = ((cur << 1) | ((byte >> bit) & 1)) >>> 0;
      length += 1;
      const bucket = table[length];
      if (bucket && Object.prototype.hasOwnProperty.call(bucket, cur)) {
        out.push(String.fromCharCode(bucket[cur]));
        cur = 0;
        length = 0;
        if (out.length === count) return out.join('');
      }
    }
  }
  return out.join('');
}

module.exports = { decode, load };
